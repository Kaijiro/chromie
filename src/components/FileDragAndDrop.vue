<template>
  <div class="container" v-cloak @drop.prevent="addFile" @dragover.prevent>
    <h2>Drag and drop a file here !</h2>
  </div>
</template>

<script lang="ts">
import {defineComponent} from "vue";
import LogFileParser from "../domain/parser/LogFileParser";

export default defineComponent({
  name: 'file-drag-and-drop',
  data() {
    return {
      files: [],
      parser: new LogFileParser()
    }
  },
  methods: {
    addFile(event: DragEvent) {
      const dataTransfer = event.dataTransfer;

      if (dataTransfer === null) {
        console.error("dataTransfer is null but it shouldn't be :(");
        return;
      }

      const files = dataTransfer.files;

      if(files.length === 0){
        console.error("The file list is empty :(");
        return;
      }

      window.chromie_ipc.send('parseRequest', (files[0] as File).path);
    }
  }
});
</script>
