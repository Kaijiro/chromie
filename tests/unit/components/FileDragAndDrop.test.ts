import {shallowMount} from "@vue/test-utils";
import FileDragAndDrop from "../../../src/components/FileDragAndDrop.vue";
import Mock = jest.Mock;

describe("FileDragAndDrop component", () => {
    let component: any;

    beforeEach(() => {
        component = shallowMount(FileDragAndDrop, {});
    });

    it(`should react when a file is dropped`, () => {
        const fileDroppedMock = jest.fn();

        component.vm.addFile = fileDroppedMock;
        component.find('.container').trigger('drop');

        expect(fileDroppedMock).toBeCalledTimes(1);
    });

    it(`should call the IPC channel to retrieve the encounters from the dropped file`, () => {
        // TODO It is actually a will to treat ONLY the first dropped file. Later, treat all file dropped.
        const IPCSendMock = window.chromie_ipc.send as Mock;

        component.find('.container').trigger('drop', {
            dataTransfer: {
                files: [{
                    name: "filename",
                    path: "/path/to/file"
                }]
            }
        });

        expect(IPCSendMock).toBeCalledTimes(1);
        expect(IPCSendMock.mock.calls[0][0]).toBe("CHROMIE_IPC_CHANNEL");
        expect(IPCSendMock.mock.calls[0][1]).toBe("/path/to/file");
    });

    it(`should display the drag and drop prompt`, () => {
        // It's currently a bit dirty to access these private attributes but this is the only way right now
        expect(component["__app"]._container).toMatchSnapshot();
    });
});
