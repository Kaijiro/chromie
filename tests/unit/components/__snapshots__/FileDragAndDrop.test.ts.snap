// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileDragAndDrop component should display the drag and drop prompt 1`] = `
<div
  data-v-app=""
>
  <div
    class="container"
  >
    <h2>
      Drag and drop a file here !
    </h2>
  </div>
</div>
`;
